pkg/ddc/alluxio/shutdown.go

```diff
@@ -179,12 +181,73 @@ func (e *AlluxioEngine) destroyWorkers(workers int32) (err error) {
 	err = e.List(context.TODO(), nodeList, &client.ListOptions{
 		LabelSelector: datasetLabels,
 	})
+
 	if err != nil {
 		return
 	}
 
+	currentWorkers = int32(len(nodeList.Items))
+	if expectedWorkers >= currentWorkers {
+		e.Log.Info("No need to scale in. Skip.")
+		return currentWorkers, nil
+	}
+
+	var nodes []corev1.Node
+	if expectedWorkers >= 0 {
+		e.Log.V(1).Info("Scale in Alluxio workers", "expectedWorkers", expectedWorkers)
+		// This is a scale in operation
+		runtimeInfo, err := e.getRuntimeInfo()
+		if err != nil {
+			e.Log.Error(err, "getRuntimeInfo when scaling in")
+			return expectedWorkers, err
+		}
+
+		if isGlobal, _ := runtimeInfo.GetFuseDeployMode(); !isGlobal {
+			// If fuses are deployed in non-global mode, workers and fuses will be scaled in together.
+			// It can be dangerous if we scale in nodes where there are pods using the related pvc.
+			// So firstly we filter out such nodes
+			pvcMountNodes, err := kubeclient.GetPvcMountNodes(e.Client, e.name, e.namespace)
+			if err != nil {
+				e.Log.Error(err, "GetPvcMountNodes when scaling in")
+				return expectedWorkers, err
+			}
+
+			for _, node := range nodeList.Items {
+				if _, found := pvcMountNodes[node.Name]; !found {
+					nodes = append(nodes, node)
+				}
+			}
+		} else {
+			// If fuses are deployed in global mode. Scaling in workers has nothing to do with fuses.
+			// All nodes with related label can be candidate nodes.
+			nodes = nodeList.Items
+		}
+
+		// Prefer to choose nodes with less data cache.
+		// Since this is just a preference, anything unexpected will be ignored.
+		worker2UsedCapacityMap, err := e.getWorkerUsedCapacity()
+		if err != nil {
+			e.Log.Info("getWorkerUsedCapacity when scaling in. Got err: %v. Ignore it", err)
+		}
+
+		if worker2UsedCapacityMap != nil && len(nodes) >= 2 {
+			// Sort candidate nodes by used capacity in ascending order
+			sort.Slice(nodes, func(i, j int) bool {
+				usageNodeA := getUsedCapacity(nodes[i], worker2UsedCapacityMap)
+				usageNodeB := getUsedCapacity(nodes[j], worker2UsedCapacityMap)
+				return usageNodeA < usageNodeB
+			})
+		}
+	} else {
+		// Destroy all workers. This is a subprocess during deletion of AlluxioRuntime
+		nodes = nodeList.Items
+	}
+
 	// 1.select the nodes
-	for _, node := range nodeList.Items {
+	for _, node := range nodes {
+		if expectedWorkers == currentWorkers {
```

Comment:
double check? Repeated with line 190.

Reply:
line 190 aims to check if this is a unnecessary scale in operation (i.e. expectedWorkers >= currentWorkers), this may happen when the scale in succeeded but  `CurrentWorkerNumberScheduled` is not successfully updated in func `SyncReplicas`.

The if statement here aims to stop the scale in until expectedWorkers is reached